src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+*/
+
+#ifndef SRC_IR_AUTH_LOGIC_UNIVERSE_RELATION_INSERTION_H_
+#define SRC_IR_AUTH_LOGIC_UNIVERSE_RELATION_INSERTION_H_
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/auth_logic_ast_traversing_visitor.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+class UniverseRelationInsertion {
+ public:
+  // A map from the names of principal / predicate argument literals
+  // (the constants wrapped in quotes) to typings.
+  using TypeEnv = absl::flat_hash_map<std::string, datalog::ArgumentType>;
+  // A map from predicate names to predicate declarations.
+  using DeclEnv =
+      absl::flat_hash_map<std::string, datalog::RelationDeclaration>;
+
+  static Program InsertUniverseRelations(const Program& prog) {
+    // Generate relation declaration environment
+    RelationDeclarationEnvironmentVisitor rel_visitor;
+    prog.Accept(rel_visitor);
+    DeclEnv decl_env = rel_visitor.decl_env();
+    // Generate constant type environment
+    TypeEnvironmentGenerationVisitor type_env_visitor(decl_env);
+    prog.Accept(type_env_visitor);
+    TypeEnv type_env = type_env_visitor.type_env();
+    // Generate universe populating facts
+    std::vector<SaysAssertion> new_says_assertions =
+        GetUniverseDefiningFacts(type_env);
+    // Transform assertions to include conditions
+    // that check the universes
+    UniverseConditionTransformer universe_condition_transformer(decl_env);
+    for (const auto& says_assertion : prog.says_assertions()) {
+      new_says_assertions.push_back(
+          universe_condition_transformer.AddUniverseConditions(says_assertion));
+    }
+
+    // Add declarations for universe relations.
+    auto universe_declarations = GetUniverseDeclarations(prog);
+    auto new_declarations = prog.relation_declarations();
+    new_declarations.insert(new_declarations.end(),
+                            universe_declarations.begin(),
+                            universe_declarations.end());
+
+    return Program(new_declarations, new_says_assertions, prog.queries());
+  }
+
+ private:
+  static std::string TypeToUniverseName(datalog::ArgumentType arg_type) {
+    switch (arg_type.kind()) {
+      case datalog::ArgumentType::Kind::kNumber:
+        return "isNumber";
+      case datalog::ArgumentType::Kind::kPrincipal:
+        return "isPrincipal";
+      case datalog::ArgumentType::Kind::kCustom:
+        return absl::StrCat("is", arg_type.name());
+    }
+  }
+
+  static BaseFact MakeUniverseMembershipFact(std::string element_name,
+                                             datalog::ArgumentType arg_type) {
+    return BaseFact(
+        datalog::Predicate(TypeToUniverseName(arg_type),  // universe relation
+                           {element_name},  // the element to add to universe
+                           datalog::Sign::kPositive));
+  }
```

Comment:
With the current functionality of this compiler pass there is no additional value in distinguishing universe membership facts from other things. I think if we see some actual functional reason to distinguish these from other kinds of facts we can implement that to meet this goal when we see this reason. Otherwise we are adding complexity to the code and spending time to implement this without gaining a benefit.